observable_operator.go

```diff
@@ -1145,6 +1145,95 @@ func (op *ignoreElementsOperator) end(_ context.Context, _ chan<- Item) {
 func (op *ignoreElementsOperator) gatherNext(_ context.Context, _ Item, _ chan<- Item, _ operatorOptions) {
 }
 
+func (o *ObservableImpl) Join(joiner Func2, right Observable, ttExtractor func(interface{}) int64, window int64, opts ...Option) Observable {
+	option := parseOptions(opts...)
+	ctx := option.buildContext()
+	next := option.buildChannel()
+
+	go func() {
+		defer close(next)
+		rBuf := []Item{}
+
+		lObserve := o.Observe()
+		rObserve := right.Observe()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case lItem, ok := <-lObserve:
+				if lItem.V == nil && !ok {
+					return
+				}
+				if lItem.Error() {
+					next <- lItem
+					return
+				}
+				lTime := ttExtractor(lItem.V)
+				cutPoint := 0
+				for i, rItem := range rBuf {
+					rTime := ttExtractor(rItem.V)
+					if (lTime <= rTime && rTime <= lTime+window) || (rTime <= lTime && lTime <= rTime+window) {
+						i, err := joiner(ctx, lItem.V, rItem.V)
+						if err != nil {
+							next <- Error(err)
+							return
+						}
+						next <- Of(i)
+					}
+					if lTime >= rTime {
+						cutPoint = i + 1
+					}
+				}
+
+				rBuf = rBuf[cutPoint:]
+
+				const timeout = time.Second
+				timeoutTimer := time.NewTimer(timeout)
```

Comment:
I don't really get what's the point of such timer.

Reply:
Fixed